scrapy_po/middleware.py

```diff
@@ -21,45 +23,44 @@ class InjectPageObjectsMiddleware:
     """
     @inlineCallbacks
     def process_response(self, request: Request, response, spider):
-        # find out which arguments need to be created
+        # find out the dependencies
         callback = get_callback(request, spider)
-        kwargs_to_provide = andi.to_provide(callback,
-                                            can_provide=self.is_page_object)
+        providers = build_providers(response)
+        can_provide = can_provide_fn(providers)
+        plan = andi.plan(callback, can_provide,
+                         providers.__contains__)
 
-        # create WebPage objects and pass them to the callback
-        for argname, cls in kwargs_to_provide.items():
-            page = yield create_page_object(cls, response)
-            request.cb_kwargs[argname] = page
+        # Build all instances declared as dependencies
+        instances = {}
+        for cls, params in plan.items():
+            if cls in providers:
+                instances[cls] = yield maybeDeferred(providers[cls])
+            elif cls == andi.FunctionArguments:
+                pass
+            else:
+                kwargs = {param: instances[pcls]
+                          for param, pcls in params.items()}
+                instances[cls] = cls(**kwargs)
```

Comment:
I think it'd be nice to move this code to a separate function, similar to andi.build or create_page_object

Reply:
It can be seen as "async andi.build", and it may be useful to have it available for use outside a middleware.

Reply:
Done in https://github.com/scrapinghub/scrapy-po/pull/5/commits/61b1b53d3015cb40d17d212bd8e67eaf031cc0a7 and https://github.com/scrapinghub/scrapy-po/pull/5/commits/7d343c9dbede0cbf495b9d022f2816f37ca2fb33